### ISSUE TITLE
Android: Role description is announced before the components text, rather than after #31042

### ISSUE THREAD

**Issue by fabOnReact** [fabriziobertoglio1987/react-native-notes#14]
Issue https://github.com/facebook/react-native/issues/31042
PR https://github.com/facebook/react-native/pull/33690

**Comment by fabOnReact** [comment]
https://github.com/fabriziobertoglio1987/react-native-notes/blob/472d531ae867ac23a866851cb51759053b267b1c/Libraries/Components/View/ViewNativeComponent.js#L22-L25

https://github.com/fabriziobertoglio1987/react-native-notes/blob/472d531ae867ac23a866851cb51759053b267b1c/ReactAndroid/src/main/java/com/facebook/react/fabric/FabricComponents.java#L26-L43

https://github.com/fabriziobertoglio1987/react-native-notes/blob/472d531ae867ac23a866851cb51759053b267b1c/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewProps.java#L19

https://github.com/fabriziobertoglio1987/react-native-notes/blob/472d531ae867ac23a866851cb51759053b267b1c/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewManager.java#L38-L40

**Comment by fabOnReact** [comment]
- [ ] Test a quick solution in java to solve this issue
- [ ] Update notes on github

**Comment by fabOnReact** [comment]
https://github.com/facebook/react-native/pull/31296#discussion_r616184584
The issues seems to be caused when passing accessibilityRole="button" to a TouchableNativeFeedback https://github.com/fabriziobertoglio1987/react-native-notes/issues/1#issuecomment-1029609228

**Comment by fabOnReact** [comment]
#### The View does not announce the child component Text when accessibilityLabel is missing (automatic content grouping)

The View has accessibilityRole="button" and child component with:
- importantForAccessibility="no". It is not focusable or clickable. The ScreenReader focus will not move on this element.
- has text `This is a child Text`

**Expected Result**:
The View announces `This is a child Text, Button`.

**Actual Result**:
The View announces `Button`.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
<View
  style={{height: 100, width: 400, backgroundColor: 'red'}}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <Text
    importantForAccessibility="no"
    style={{
      height: 50,
      marginLeft: 50,
      marginTop: 50,
      backgroundColor: 'white',
    }}>
    {childText}
  </Text>
</View>
```
</p>
</details>

<details><summary>click to open recording of test case</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/161759566-01a9da7c-b367-4a46-82dd-e32ef8570343.mp4" width="1000" />

</p>
</details>

**Comment by fabOnReact** [comment]
#### The View announces the child component Text when accessibilityLabel is missing (automatic content grouping)

The View has accessibilityRole="button" and child component with:
- importantForAccessibility="no". It is not focusable or clickable. The ScreenReader focus will not move on this element.
- has text `This is a child Text`

**Expected/Actual Result**:
The View announces `This is a child Text, Button`.

**Notes**
- The solution here is checking if an accessibilityLabel already exist otherwise over-ride it with the childText
- The child text should be non focusable, otherwise their label/text is announced when TalkBack ScreenReader focuses on the element
- The fix is applied to the affected components (Button, TouchableOpacity etc..)

<details><summary>click to open code used in the test</summary>
<p>

```javascript
<View
  style={{height: 100, width: 400, backgroundColor: 'red'}}
  importantForAccessibility="yes"
  accessibilityRole="button"
  accessibilityLabel={props.accessibilityLabel ? props.accessibilityLabel : childText}>
  <Text
    importantForAccessibility="no"
    style={{
      height: 50,
      marginLeft: 50,
      marginTop: 50,
      backgroundColor: 'white',
    }}>
    {childText}
  </Text>
</View>
```
</p>
</details>

<details><summary>click to open recording of test case</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/161760880-890b0936-434c-42ce-907c-19010f2a80af.mp4" width="1000" />

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
onInitializeAccessibilityNodeInfo is called after onChangeText. 
When we change the text, nodeInfo.setError() is triggered, but the error displays on the next onChangeTextEvent (the next typed char). We whould trigger accessibility announcement after calling setError().

This issue does not reproduce when we use editText.setError().
`setError` calls a method [notifyViewAccessibilityStateChangedIfNeeded][1], which may be responsible to trigger the accessibility announcement

[1]: https://github.com/aosp-mirror/platform_frameworks_base/blob/1ac46f932ef88a8f96d652580d8105e361ffc842/core/java/android/widget/TextView.java#L7284 "notifyViewAccessibilityStateChangedIfNeeded"

- [ ] try to invoke the accessibility announcement manually

**Comment by fabriziobertoglio1987** [comment]
- [ ] caching

**Comment by fabOnReact** [comment]
Scenarios - TouchableNativeFeedback

<details><summary>Touchable with one child</summary>
<p>

```
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <Text
    accessible={false}
    focusable={false}>
    Text number 1
  </Text>
</TouchableNativeFeedback>
```

</p>
</details>

<details><summary>Touchable with multiple children</summary>
<p>

```
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <View>
      <Text
        accessible={false}
        focusable={false}>
        Text number 1
      </Text>
      <Text
        accessible={false}
        focusable={false}>
        Text number 2
      </Text>
   </View>
</TouchableNativeFeedback>
```

</p>
</details>


<details><summary>Touchable with children that have children</summary>
<p>

```
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <View>
      <Text
        accessible={false}
        focusable={false}>
        Text number 1
      </Text>
      <Text
        accessible={false}
        focusable={false}>
        Text number 2
          <Text
            accessible={false}
            focusable={false}>
            Text number 3
          </Text>
      </Text>
   </View>
</TouchableNativeFeedback>
```

</p>
</details>

<details><summary>Touchable with children's that are not Text components</summary>
<p>

```
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <View>
      <Text
        accessible={false}
        focusable={false}>
        Text number 1
      </Text>
      <Text
        accessible={false}
        focusable={false}>
        Text number 2
      </Text>
   </View>
</TouchableNativeFeedback>
```

</p>
</details>

<details><summary>Touchable with one Text link</summary>
<p>

```
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">          
    <Text accessible={false} focusable={false}>
      Text number 1<Text accessibilityRole="link">Link</Text>
    </Text>
</TouchableNativeFeedback>
```

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
- [ ] write a post where you break down this issue (current solution needs to handle all scenarios and scale)

**Comment by fabriziobertoglio1987** [comment]
First Issue:
- isAccessibleFocusable is never true
- we are not looping through each child and adding the text
Second Issue:
- the problem now is that childNodeDescription isEmpty
- it is empty because getTalkbackDescription does not return the text of the Text child

Chainging getTalkbackDescription to just return node.getText() fixes the issue.

**Comment by fabOnReact** [comment]
- [x] isAccessibilityFocused
- [x] isSpeakingNode

**Comment by fabriziobertoglio1987** [comment]
The current issue with the implementation is:
- isAccessibilityFocused does not return the correct value from onInitialializeAccessibilityNodeInfo
- isSpeakingNode returns the wrong value

**Comment by fabOnReact** [comment]
- [x] add check if parent node already has accessibilityLabel or contentDescription to be read, if not then retrieve children
- [x] scenario where getTalkbackDescription returns null
- [x] add grid role as included in https://github.com/facebook/react-native/pull/33180
- [x] add test cases to AccessibilityExample
- [x] save test cases for button, view and other components
- [x] example where text has accessibilityAction or is clickable etc..
- [x] verify no regression in [ScrollView](https://github.com/fabriziobertoglio1987/react-native-notes/issues/14#issuecomment-1106012823)
- [x] rebase
- [x] update pr summary
- [x] instead of using AccessibilityNodeInfoCompat.obtain() use node.getChild to loop over children nod
- [ ] test more edge cases

Low Priority:

- [ ] implement [getRole()](https://github.com/facebook/flipper/blob/e79b0fd96fae8bbe4fe0b05e827ace3e96580a4b/android/src/main/java/com/facebook/flipper/plugins/inspector/descriptors/utils/AccessibilityRoleUtil.java#L116) instead of check on [accessibilityRole variable](https://github.com/fabriziobertoglio1987/react-native/blob/a05b912d9dab1a341e8d6ec234f27394cc7f0f2d/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L313)

[1]: https://github.com/fabriziobertoglio1987/react-native/blob/bdabc4c825d6464e86a600054a28b75eed2cf895/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L506

<details><summary>completed</summary>
<p>

- [x] test edge cases
- [x] evaluate if we need [isTopLevelScrollItem][1] 
- [x] implement logic to only handle the basic scenario
      implement logic for isTopLevelScrollItem
- [x] compare current implementation with Flipper
- [x] prepare draft implementation
- [x] evaluate adding/removing code after comparing with Flipper
- [x] test EditText and verify behaviour is correct
- [x] prepare pr summary

</p>
</details>

<details><summary>low priority</summary>
<p>

- [x] getRole to avoid triggering runtime error
- [x] scenario - scroll view or other components
- [ ] adding space when several text components next to each other
- [x] consider adding a check on the accessibilityRole=button
      do we need this behaviour only for the items with accessibilityRole button?
      if yes, then we add a check on the accessibilityRole

</p>
</details>

**Comment by fabOnReact** [comment]
#### Screenreader focuses on ScrollView Items

**Expected/Actual Result**:
This has to be verified again before moving the pr to Ready for Review.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function TouchableExample(props) {
  return (
    <ScrollView
      accessible={true}
      focusable={true}
      style={{backgroundColor: 'red'}}>
      <Text>Text number 1</Text>
      <Text>
        Text number 2<Text accessible={false}>Text number 3</Text>
      </Text>
    </ScrollView>
  );
}
```
</p>
</details>

<details><summary>click to open recording of test case - without the implementation</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165452565-948ba012-2e77-4b99-b4a9-8adbd9ca99cf.mp4" width="1000" />

</p>
</details>

<details><summary>click to open recording of test case - with the implementation</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/164606571-36080e19-1b01-43a7-a872-ab04246db667.mp4" width="1000" />

</p>
</details>

**Comment by fabriziobertoglio1987** [comment]
#### Pressable does not announce accessibilityRole

Screenreader focus moves to the to the Pressable component:

**Expected Result**:
The screenreader announces "Text number 1, Text number 2Text number 3, button"

**Actual Result**:
**without accessibilityRole**: The screenreader announces "Text number 1, Text number 2Text number 3"
**with accessibilityRole**: The screenreader announces "button"

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function TouchableExample(props) {
  return (
    <Pressable
      accessible={true}
      focusable={true}
      onPress={() => console.warn('pressed')}
      style={{backgroundColor: 'red'}}>
      <Text accessible={false}>Text number 1</Text>
      <Text accessible={false}>
        Text number 2<Text accessible={false}>Text number 3</Text>
      </Text>
    </Pressable>
  );
}
```
</p>
</details>

<details><summary>click to open recording of test case</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/164607421-8e20e539-985a-44ee-a13f-1e03dd7d56b4.mp4" width="1000" />

</p>
</details>

**Comment by fabOnReact** [comment]
#### Screenreader announces Pressability items

**Expected/Actual Result**:
Screenreader announces `Text number 1Text number 2 Text number 3, Button`.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function TouchableExample(props) {
  return (
    <Pressable
      accessible={true}
      focusable={true}
      onPress={() => console.warn('pressed')}
      accessibilityRole="button"
      style={{backgroundColor: 'red'}}>
      <View>
        <Text accessible={false}>Text number 1</Text>
        <Text accessible={false}>
          Text number 2<Text accessible={false}>Text number 3</Text>
        </Text>
      </View>
    </Pressable>
  );
}
```
</p>
</details>

<details><summary>click to open recording of test case</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/164671364-587676ad-1a57-4fb5-8d8a-1bfd4ebe6e88.mp4" width="1000" />



</p>
</details>

**Comment by fabOnReact** [comment]
#### Screenreader correctly announcing accessible/non-accessible items

**Expected/Actual Result**:
Talkback announces `Text number 1Text number 2Text number3, Button`.
Talkback announces `Text clickable`.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function TouchableExample(props) {
  return (
    <Pressable
      accessible={true}
      focusable={true}
      onPress={() => console.warn('pressed')}
      accessibilityRole="button"
      style={{backgroundColor: 'red'}}>
      <View>
        <Text onPress={() => console.warn('pressed')}>Text clickable</Text>
        <Text accessible={false}>Text number 1</Text>
        <Text accessible={false}>
          Text number 2<Text accessible={false}>Text number 3</Text>
        </Text>
      </View>
    </Pressable>
  );
}
```
</p>
</details>

<details><summary>click to open recording of test case</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/164700912-25a39bd4-40a0-4491-bcaa-623e3f4e605f.mp4" width="1000" />


</p>
</details>

**Comment by fabOnReact** [comment]
#### Conditions that should trigger the custom contentDescription

https://github.com/facebook/react-native/pull/33690#issuecomment-1106779940

The custom contentDescription should be triggered if:
- an element has no contentDescription **AND** no text 
- **AND** has other content to announce (role, state, etc.)

**Comment by fabOnReact** [comment]
#### Runtime error when calling createNodeInfoFromView 

```java
  @Nullable
  public static AccessibilityNodeInfoCompat createNodeInfoFromView(View view) {
    if (view == null) {
      return null;
    }

    final AccessibilityNodeInfoCompat nodeInfo = AccessibilityNodeInfoCompat.obtain();

    // For some unknown reason, Android seems to occasionally throw a NPE from
    // onInitializeAccessibilityNodeInfo.
    try {
      // ==> Runtime when called from onInitializeAccessbilityNodeInfo
      ViewCompat.onInitializeAccessibilityNodeInfo(view, nodeInfo);
    } // ...
}
```
create an infinite loop, `onInitializeAccessibilityNodeInfo` calls method that calls itself without stopping.

**Comment by fabOnReact** [comment]
#### Button role is announced after child contentDescription with TouchableNativeFeedback

**Expected/Actual Result**:
Talkback announces `Text number 1, Text number 1, Button`

<details><summary>click to open code used in the test</summary>
<p>

```javascript
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <View accessible={false}>
    <Text accessible={false} focusable={false}>
      Text number 1
    </Text>
    <Text accessible={false} focusable={false}>
      Text number 1
    </Text>
  </View>
</TouchableNativeFeedback>
```
</p>
</details>

<details><summary>without fix - Button role is announced before child text</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165335974-d45e96b4-acfe-440f-822d-93a973de3de3.mp4" width="1000" />

</p>
</details>

<details><summary>without fix - Button role is announced before child text (second example)</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165436867-68abfc34-a904-4c6a-920a-a19c2a55124c.mp4" width="1000" />

</p>
</details>

<details><summary>with fix - Button role is announced after child text</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165336588-23eb801c-79a7-4e1c-b1bf-85c1d925b973.mp4" width="1000" />

</p>
</details>

<details><summary>with fix - Button role is announced after child text (second example)</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165436943-2aa8621c-6454-4b47-ae07-33130a4d0425.mp4" width="1000" />

</p>
</details>

**Comment by fabOnReact** [comment]
#### Testing for regressions in Accessibility Actions

**Expected/Actual Result**: Same behavior with/out implementation

<details><summary>click to open code used in the test</summary>
<p>

```javascript
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button"
  accessibilityActions={[
    {name: 'activate', label: 'activate label'},
    {name: 'copy', label: 'copy label'},
  ]}
  onAccessibilityAction={event => {
    switch (event.nativeEvent.actionName) {
      case 'activate':
        Alert.alert('Alert', 'Activate accessiblity action');
        break;
      case 'copy':
        Alert.alert('Alert', 'copy action success');
        break;
    }
  }}>
  <View>
    <Text accessible={false}>Text number 1</Text>
    <Text accessible={false}>
      Text number 2<Text accessible={false}>Text number 3</Text>
    </Text>
  </View>
</TouchableNativeFeedback>
```
</p>
</details>

<details><summary>click to open recording of test case - without implementation</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165443634-f987d1a6-7ef6-4808-a2a3-96ae7542d20a.mp4" width="1000" />

</p>
</details>

<details><summary>click to open recording of test case - with implementation</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165443724-897971f0-b1e8-495e-92f8-fa01adfd72bd.mp4" width="1000" />

</p>
</details>

**Comment by fabOnReact** [comment]
- [x] Check this condition works when calculating child text 
>Talkback defines as having clickable=true, longClickable=true, or focusable=true, or having AccessibilityActions for any of those)
- [ ] Add more edge cases

**Comment by fabOnReact** [comment]
#### Recordings of complete test cases in rn-tester app main and pr branch

<details><summary>click to open recording of test case - main branch</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165717229-99ff5605-e9e3-4814-a10a-89fe762e6406.mp4" width="1000" />

</p>
</details>

<details><summary>click to open recording of test case - pr branch</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/165718512-9be75adf-9015-418b-8ca9-2805d85efe18.mp4" width="1000" />



</p>
</details>

**Comment by fabOnReact** [comment]
- [ ] ~import AccessibiltyNodeInfoUtils instead of replicating logic. they are static methods~
       ~Try to use logic from Flipper AcceUtil instead of copy paste each method.~ 
- [ ] ~use [AccessibilityRole.from/toValue](https://github.com/fabriziobertoglio1987/react-native/blob/9d1bf23028ef056ae0aba4283e727b2b4e239f52/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java#L175)~
- [x] fix circle ci test cases (try to do rebase)

**Comment by fabOnReact** [comment]
<details><summary>TouchableOpacity with child EditText annouces Button role before the child contentDescription</summary>
<p>

test case added with commit https://github.com/fabriziobertoglio1987/react-native/commit/9d72a70c3b27d8fb198dd32a1cd8c031e458d089

The issue is caused by this [line](https://github.com/fabriziobertoglio1987/react-native/blob/9d72a70c3b27d8fb198dd32a1cd8c031e458d089/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L308). `info.getText()` retrieves the text of the child TextInput (the placeholder), but the current element (in js the TouchableNativeFeedback), does not have contentDescription or Text and the role (button) of the TouchableNativeFeedback is announced before the contentDescription of the child component.

TouchableNativeFeedback does not exist in ReactAndroid, and it is just a wrapper to return the child component with an onPress event handler, for this reason `info.getText()` returns the child text.

The functionalities are developed with https://github.com/fabriziobertoglio1987/react-native/commit/9d1bf23028ef056ae0aba4283e727b2b4e239f52 

<details><summary>test done on the 27.05</summary>
<p>

in this case the first text is not announced, because it's a secure entry.

<video src="https://user-images.githubusercontent.com/24992535/170686887-34c515b6-2d85-481b-b42b-0a898d0aa0a1.mp4" width="200" />

</p>
</details>

<details><summary>test done on the 30.05</summary>
<p>

on commit https://github.com/fabriziobertoglio1987/react-native/commit/abed89ec86affa80f41a03eaf14e0cf41f1f1dbf

<video src="https://user-images.githubusercontent.com/24992535/170980322-99df0839-c45a-48e4-a3c4-799ee42c9ad7.mp4" width="200" />

</p>
</details>


</p>
</details>

**Comment by fabOnReact** [comment]
<details><summary>TouchableOpacity with child EditText annouces placeholder text before and after the role</summary>
<p>

The issue reproduces on the main/pr branch.
The placeholder text is announced before and after the role.

**Expected Result** this is the placeholder, Button
**Actual Result** this is the placeholder, Button, this is the placeholder

[Explanation](https://user-images.githubusercontent.com/24992535/170980322-99df0839-c45a-48e4-a3c4-799ee42c9ad7.mp4) of the result:
- The first announcement corresponds to the value returned from `nodeInfo.getText()`
- The second announcement corresponds to the role
- The last announcement corresponds to the hint text

Flipper includes implementation of [getTalkbackHint](https://github.com/facebook/flipper/blob/e44cad5e9912b359378a43ec47c5a49bbca85fc3/android/src/main/java/com/facebook/flipper/plugins/inspector/descriptors/utils/AccessibilityUtil.java#L376) and ([getTalkbackData](https://github.com/facebook/flipper/blob/e44cad5e9912b359378a43ec47c5a49bbca85fc3/android/src/main/java/com/facebook/flipper/plugins/inspector/descriptors/utils/AccessibilityUtil.java#L715-L719)).

<details><summary>Source Code of Test Case</summary>
<p>

```javascript
<TouchableNativeFeedback
  accessible={true}
  importantForAccessibility="yes"
  accessibilityRole="button">
  <TextInput
    accessible={false}
    style={styles.default}
    placeholder="this is the placeholder"
  />
</TouchableNativeFeedback>
```

</p>
</details>

<details><summary>Test Case of Main Branch</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/170971112-3d30b294-682b-4e66-a030-aa4b54467f1a.mp4" with="200"></video>

</p>
</details>

<details><summary>Test Case of PR Branch</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/170970366-ce8b1e70-baac-4ce1-98d9-0e0fd3984c24.mp4" with="200"></video>

</p>
</details>

</p>
</details>

**Comment by fabOnReact** [comment]
<details><summary>TouchableOpacity with TextInput child announces contentDescription before the Role</summary>
<p>


<video src="https://user-images.githubusercontent.com/24992535/170975845-c8497f82-49ea-4d96-b922-426edf900813.mp4" width="200" />


</p>
</details>

**Comment by fabOnReact** [comment]
<details><summary>EditText is allways accessible to TalkBack screenreader</summary>
<p>

I could not find a test scenarios for this [logic](https://github.com/fabriziobertoglio1987/react-native/blob/388911e4e7b27a627def6e276562018e7e400c4f/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L1053-L1075) implemented with commit https://github.com/fabriziobertoglio1987/react-native/commit/388911e4e7b27a627def6e276562018e7e400c4f 

The EditText is always accessible, so it makes no sense to over-ride the TalkBack logic.
The logic could be deleted.

<details><summary>Sourcecode</summary>
<p>

The [code](https://github.com/fabriziobertoglio1987/react-native/blob/388911e4e7b27a627def6e276562018e7e400c4f/packages/rn-tester/js/examples/Accessibility/AccessibilityExample.js#L283-L299) of the example

```javascript
        <RNTesterBlock title="TouchableNativeFeedback with non-accessible child Texts, one of them is an EditText">
          <View
            accessible={true}
            importantForAccessibility="yes"
            accessibilityRole="button">
            <View>
              <Text accessible={false}>Text number 1</Text>
              <TouchableNativeFeedback accessible={false} focusable={false}>
                <TextInput
                  accessible={false}
                  focusable={false}
                  style={styles.default}
                />
              </TouchableNativeFeedback>
            </View>
          </View>
        </RNTesterBlock>
```

</p>
</details>




https://user-images.githubusercontent.com/24992535/171169545-339dc638-3938-4166-928a-bba47b280a75.mp4



</p>
</details>

**Comment by fabOnReact** [comment]
The default implementation of AccNodeInfo.getRoleDescription does not work with all roles:

1) image => “Text number 1, Image, Text number 2, Image” 
   With a parent role button => “Text number 1, Text number 2, Button”
2) text => “Text number 1, Text number 2”
3) button => the component with role button becomes accessible
4) search => “Text number 1, EditBox, Text number 2”

- [ ] test custom implementation of getRoleDescription
- [ ] handle scenario in custom implementation when role not exist and getRoleDescription returns the class of the child
        for example: `<Text>example</Text>` will add role Text to the component, because the role is retrieved from the Java class.
        check Talkback Java implementation of getRoleDescription
- [ ] add correct capitalization to custom implementation

**Comment by fabOnReact** [comment]
- [ ] code reuse - use talkback [isSpeakingNode](https://github.com/google/talkback/blob/6c0b475b7f52469e309e51bfcc13de58f18176ff/utils/src/main/java/com/google/android/accessibility/utils/AccessibilityNodeInfoUtils.java#L905) instead of making new [implementation](https://github.com/fabriziobertoglio1987/react-native/blob/ad6732aa3576786d37478a729b112031dd94b682/ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactAccessibilityDelegate.java#L827). 
This method can call itself recursively through #hasNonActionableSpeakingChildren

**Comment by fabOnReact** [comment]
<details><summary>One of the child has accessibilityState (hasStateDescription triggers the announcement)</summary>
<p>

https://github.com/fabriziobertoglio1987/react-native/commit/c17f00ea321221d379a4b9e4fefbca991f50afc1
https://github.com/facebook/react-native/pull/33690#discussion_r890259702

<details><summary>sourcecode from test case</summary>
<p>

```javascript
<View accessible={true} accessibilityRole="button">
  <Text accessible={false}>Text number 1</Text>
  <Text
    style={styles.smallRedSquare}
    accessible={false}
    accessibilityState={{checked: true}}
    accessibilityLabel="this child Text does not have text, but has state and should be announced by TalkBack"
    accessibilityRole="image"
  />
</View>
```


</p>
</details>

<video src="https://user-images.githubusercontent.com/24992535/172551047-0f90562f-4f3f-4bc0-bb52-9b1ff1f42bcd.mp4" width="200" />

</p>
</details>

**Comment by fabOnReact** [comment]
<details><summary>One of the child has accessibilityHint (hasText triggers the announcement)</summary>
<p>

https://github.com/fabriziobertoglio1987/react-native/commit/c17f00ea321221d379a4b9e4fefbca991f50afc1
https://github.com/facebook/react-native/pull/33690#discussion_r890259702

<details><summary>sourcecode from test case</summary>
<p>

```javascript
<View accessible={true} accessibilityRole="button">
  <Text
    style={styles.smallRedSquare}
    accessible={false}
    accessibilityHint="this child Text does not have text, but has hint and should be announced by TalkBack"
  />
</View>
```


</p>
</details>

<video src="https://user-images.githubusercontent.com/24992535/172551707-d6e07752-0b10-4bbe-8b37-c6f54f5816d1.mp4" width="200" />

</p>
</details>
